### PR TITLE
New test setup

### PR DESCRIPTION
**Pull request by arunoda**
Implements https://github.com/zeit/next.js/issues/635

This PR comes with a bunch of improvements for the test setup. This will allow us to write test faster and better. Here are the things comes with this PR.

* Move testing into standard `jest-cli` and remove it from gulp.
* Add some isolated tests for the `server/resolve`
* Allow current integration tests to get html content via HTTP
* Use cheerio for parsing HTML in above tests
* Add a set of utils, which can be imported from anywhere with `import from 'next-test-utils'`
* Build a production app and check for the gzip usage
* Use webdriver to test client side navigations

Earlier, I thought about writing integration tests in the example directory. (Inside examples). But those test apps might be optimized for tests. So, they are not exactly example apps.

We've almost everything needed for a better test suite. After we take this PR, we can write more tests including tests for HMR.

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -52,6 +52,15 @@ export default class Server {
     if (this.hotReloader) {
       await this.hotReloader.stop()
     }
+
+    if (this.http) {
```
Comment (arunoda) [inline server/index.js]: We need this because, we need to close the HTTP server after each test suite.

File: package.json [hunk 2]
```diff
@@ -23,11 +23,10 @@
   },
   "scripts": {
     "build": "gulp",
-    "pretest": "npm run lint",
-    "test": "gulp test",
-    "html-report": "nyc report --temp-directory=./coverage --reporter=html",
-    "jest": "jest --coverage",
-    "coveralls": "nyc report --temp-directory=./coverage --reporter=text-lcov | coveralls",
```
Comment (arunoda) [inline package.json]: How does coveralls works. Do we need this?

**Comment by bcoe** [comment]
@arunoda this seems to overlap some of the work I've been doing with a new helper called `next-test-helper`:

https://github.com/zeit/next.js/pull/629

@rauchg would it be possible to add @arunoda to this repo?

File: package.json (hunk 2 above)
Comment (bcoe) [inline package.json]: @arunoda coveralls needs to be piped the `lcov` format test output post test:

https://github.com/zeit/next.js/blob/master/.travis.yml#L14

so this logic is necessary.

Comment (arunoda) [inline package.json]: Oh! Thanks.
I'll add it back.

**Comment by arunoda** [comment]
@bcoe Do you know how to add source map support to coverage reports?
If that's possible, that's superb.

**Comment by bcoe** [comment]
@arunoda Jest should already be applying source-maps appropriately, source-maps are also built into `nyc` itself; although `nyc` isn't currently performing the instrumentation, this is all being done by `babel-plugin-istanbul` which is part of Jest:

https://github.com/facebook/jest/pull/2290

**Comment by bcoe** [comment]
@arunoda as far as I understand it, Jest should already be running everything through `babel-plugin-istanbul`; so I don't think you should need to worry about adding that plugin, unless I'm missing something.

**Comment by arunoda** [comment]
@bcoe Really. I'll check.

**Comment by arunoda** [comment]
@bcoe I think that's okay if we give Jest ES6 code. But currently, we give it transpiled code. That's by running the babel transformation on our own (in pretest).

So, I think that's needed for now.

**Comment by arunoda** [comment]
@bcoe I could get the source-maps working. But it shows reports from `dist` as well.
See: https://coveralls.io/builds/9510584
(That's how it show on the command line Jest report as well)

I just wanna ignore `dist`. I tried different Jest options. But didn't work.
Any help will be awesome :)

**Comment by rauchg** [comment]
@cpojer

File: package.json [hunk 3]
```diff
@@ -23,11 +23,10 @@
...
+    "pretest": "npm run lint && NODE_ENV=test npm run prepublish",
+    "test": "NODE_PATH=test/lib jest test/unit/*.test.js test/isolated/*.test.js test/integration/**/test/*.test.js --coverage",
+    "testonly": "NODE_PATH=test/lib jest test/unit/*.test.js test/isolated/*.test.js test/integration/**/test/*.test.js",
+    "coveralls": "nyc --instrument=false --source-map=false report --temp-directory=./coverage --reporter=text-lcov | coveralls",
```
Comment (cpojer) [inline package.json]: we use `codecov` for this on Jest itself and don't really need nyc in this case.

**Comment by cpojer** [comment]
Not really sure why I was pulled into this other than cheering on you. Thanks for doing this and simplifying the setup.

**Comment by bcoe** [comment]
@cpojer @arunoda don't have too much time to dig into this today, however!

A couple notes:

* Jest has the same coverage internals built into it that nyc is using; you should be able to leave the coveralls reporting as: `nyc report --temp-directory=./coverage --reporter=text-lcov | coveralls`:
  * Jest will output the coverage results into `./coverage`, youre then simply reloading the output and piping it in `lcov` format to coveralls,
* limiting the files that are instrumented for coverage needs to happen during the test run, not post-hoc when you report the data to coveralls; to do this you can use: https://facebook.github.io/jest/docs/configuration.html#coveragepathignorepatterns-array-string

File: package.json (hunk 3 above)
Comment (arunoda) [inline package.json]: Got it.

**Comment by arunoda** [comment]
@bcoe @cpojer Thanks for ideas and suggestions.

I had no luck. But anyway, this is a minor issue and we can fix this later on.
@rauchg we only need few things left to make this setup looks good. I'll do it them today.

**Comment by arunoda** [comment]
@nkzawa This PR is ready. Could you review this?

File: package.json [hunk 4]
```diff
@@ -23,11 +23,12 @@
   },
   "scripts": {
     "build": "gulp",
-    "pretest": "npm run lint",
-    "test": "gulp test",
-    "html-report": "nyc report --temp-directory=./coverage --reporter=html",
-    "jest": "jest --coverage",
-    "coveralls": "nyc report --temp-directory=./coverage --reporter=text-lcov | coveralls",
+    "pretestonly": "chromedriver &",
```
Comment (nkzawa) [inline package.json]: I prefer to use headless browsers for local tests if possible. I wonder if the current version of phantomjs is stable enough or can use other one.

File: package.json [hunk 5]
```diff
@@ -23,11 +23,12 @@
...
+    "testonly": "NODE_PATH=test/lib jest test/unit/*.test.js test/isolated/*.test.js test/integration/**/test/*.test.js",
+    "posttestonly": "pkill chromedriver",
+    "pretest": "npm run lint && NODE_ENV=test npm run prepublish",
+    "test": "npm run testonly -- --coverage --forceExit",
+    "coveralls": "nyc --instrument=false --source-map=false report --temp-directory=./coverage --reporter=text-lcov | coveralls",
     "lint": "standard 'bin/*' 'client/**/*.js' 'examples/**/*.js' 'lib/**/*.js' 'pages/**/*.js' 'server/**/*.js' 'test/**/*.js'",
     "prepublish": "gulp release",
```
Comment (nkzawa) [inline package.json]: `npm run prepublish` is used to run tests. It looks this is not true now ?

File: package.json (hunk 4 above)
Comment (arunoda) [inline package.json]: I tried. Due to some reason it didn't work out. It doesn't run our JS code for some reason. Debugging that is hard.

Anyway, chromedriver works pretty well and it's pretty fast.
(BTW: chromedriver is installed via NPM, so no need to setup anything)

File: package.json (hunk 5 above)
Comment (arunoda) [inline package.json]: We can do that.

File: package.json (hunk 4 above)
Comment (arunoda) [inline package.json]: BTW: chromedriver is fast as phantomjs.

File: package.json (hunk 5 above)
Comment (arunoda) [inline package.json]: Done.

**Comment by rauchg** [comment]
I'll have @nkzawa make a final decision on this

File: .travis.yml [hunk 6]
```diff
@@ -1,14 +1,23 @@
...
+dist: trusty
+addons:
+  apt:
+    sources:
+      - google-chrome
+    packages:
+      - google-chrome-stable
 language: node_js
 node_js:
-  - "node"
```
Comment (nkzawa) [inline .travis.yml]: any reasons to remove this ?

File: .travis.yml [hunk 7]
```diff
@@ -1,14 +1,23 @@
...
 cache:
-  yarn: true
   directories:
     - node_modules
-install:
-  - curl -o- -L https://yarnpkg.com/install.sh | bash
-  - export PATH=$HOME/.yarn/bin:$PATH
-  - yarn --ignore-engines
+before_install:
+  - rm yarn.lock
```
Comment (nkzawa) [inline .travis.yml]: why ? `npm` is faster in this case ?

File: gulpfile.js [hunk 8]
```diff
@@ -8,8 +8,9 @@ const sequence = require('run-sequence')
 const webpack = require('webpack')
 const webpackStream = require('webpack-stream')
 const del = require('del')
-const jest = require('gulp-jest')
+const child_process = require('child_process')
```
Comment (nkzawa) [inline gulpfile.js]: this should be `childProcess`

File: package.json [hunk 9]
```diff
@@ -54,7 +56,6 @@
     "babel-preset-react": "6.16.0",
     "babel-runtime": "6.20.0",
     "case-sensitive-paths-webpack-plugin": "1.1.4",
-    "cross-spawn": "5.0.1",
```
Comment (nkzawa) [inline package.json]: this is used on `bin/next`

File: .travis.yml (hunk 6 above)
Comment (arunoda) [inline .travis.yml]: Not particular reason. We already testing for node 4 and 6.
`node` is the same as `6`.
That's why I removed that.

File: .travis.yml (hunk 7 above)
Comment (arunoda) [inline .travis.yml]: Here we install chromedriver with NPM.
It doesn't seems to work properly with yarn + travis-ci's caching setup.

Anyway, moving to NPM back didn't add any considerable amount of time to the build.

File: package.json (hunk 9 above)
Comment (arunoda) [inline package.json]: Oh!

**Comment by nkzawa** [comment]
@arunoda can you fix conflicts ?

**Comment by arunoda** [comment]
@nkzawa sure. on it.

**Comment by arunoda** [comment]
@nkzawa done.

**Comment by nkzawa** [comment]
👍 👍 👍

**Pull request merged**